qiskit/algorithms/minimum_eigen_solvers/adapt_vqe.py

```diff
@@ -0,0 +1,336 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A ground state calculation employing the AdaptVQE algorithm."""
+from __future__ import annotations
+from typing import Optional, List, Tuple, Union
+
+import re
+import logging
+
+from enum import Enum
+import numpy as np
+from qiskit import QiskitError
+
+from qiskit.algorithms import VQE
+from qiskit.algorithms.list_or_dict import ListOrDict
+from qiskit.algorithms.minimum_eigen_solvers.vqe import VQEResult
+from qiskit.algorithms import VariationalAlgorithm
+from qiskit.circuit import QuantumCircuit
+from qiskit.opflow import OperatorBase, PauliSumOp, CircuitSampler
+from qiskit.opflow.gradients import GradientBase, Gradient
+from qiskit.circuit.library import EvolvedOperatorAnsatz
+from qiskit.utils.validation import validate_min
+from qiskit.algorithms.aux_ops_evaluator import eval_observables
+from ..exceptions import AlgorithmError
+
+
+logger = logging.getLogger(__name__)
+
+
+class FinishingCriterion(Enum):
+    """The FinishingCriterion Class is an Enum for finishing criteria."""
+
+    CONVERGED = "Threshold converged"
+    CYCLICITY = "Aborted due to a cyclic selection of evolution operators"
+    MAXIMUM = "Maximum number of iterations reached"
+    UNKNOWN = "The algorithm finished due to an unforeseen reason"
+
+
+class AdaptVQE(VariationalAlgorithm):
+    """The Adaptive Variational Quantum Eigensolver algorithm.
+
+    `AdaptVQE <https://arxiv.org/abs/1812.11173>`__ is a quantum algorithm creates a compact ansatz
+    by gradually building up the ansatz circuit by appending the excitation with the largest energy
+    gradient to the circuit. This results in a wavefunction ansatz that is uniquely formed by the
+    algorithm.
+    This is an adaptive wrapper of a VQE used internally as solver.
+    The performance of AdaptVQE can significantly depend on the choice of gradient method, QFI
+    solver (if applicable) and the epsilon value.
+    """
+
+    def __init__(
+        self,
+        solver: VQE,
+        adapt_gradient: Optional[GradientBase] = None,
+        excitation_pool: List[Union[OperatorBase, QuantumCircuit]] = None,
+        threshold: float = 1e-5,
+        max_iterations: Optional[int] = None,
+    ) -> None:
+        """
+        Args:
+            solver: a factory for the VQE solver employing a UCCSD ansatz.
+            adapt_gradient: a class that converts operator expression to the first-order gradient based
+                on the method mentioned.
```

Comment:
```suggestion
            adapt_gradient: a `Gradient` instance with which to compute the excitation operator gradients.
                If `None`, this will default to a parameter shift gradient.
```